### PR TITLE
Add basic CSC and CSR sparse matrix support

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR adds basic type support for CSC and CSR sparse matrices.  It's just a start to adding real sparse matrix support (i.e. #29), but it at least allows one to use object mode in a more constrained way and reference simple properties of sparse matrices.  More useful additions will follow.

### CODE REVIEW

**Comment by brandonwillard** [comment]
It looks like [the environment is missing a command used by the build script](https://github.com/numba/numba-scipy/actions/runs/3659744406/jobs/6186089522#step:6:238)?

**Comment by esc** [comment]
> It looks like [the environment is missing a command used by the build script](https://github.com/numba/numba-scipy/actions/runs/3659744406/jobs/6186089522#step:6:238)?

It looks like `catchsegv` isn't installed on the host doing the testing?

**Comment by guilhermeleobas** [comment]
`catchsegv` is available from [glibc](https://command-not-found.com/catchsegv). Could we get that installed? Or perhaps don't use `catchsegv` if it is not installed:

```bash
if [[ "$unamestr" == 'Linux' ]]; then
    if [[ "${BITS32}" == "yes" ]]; then
        SEGVCATCH=""
    else
        SEGVCATCH= [[ -n $catchsegv ]] || catchsegv=""
    fi
```

**Comment by guilhermeleobas** [comment]
All tests pass locally, but I had to manually import `sparse` at the top level of `test_sparse.py`

```
$ pytest -rs -sv numba_scipy/tests/test_sparse.py --tb=short -x
========================================================================================================================= test session starts ==========================================================================================================================
platform darwin -- Python 3.10.8, pytest-7.2.0, pluggy-1.0.0 -- /Users/guilhermeleobas/miniconda3/envs/numba-scipy/bin/python3.10
cachedir: .pytest_cache
rootdir: /Users/guilhermeleobas/git/numba-scipy
collected 5 items

numba_scipy/tests/test_sparse.py::test_sparse_unboxing PASSED
numba_scipy/tests/test_sparse.py::test_sparse_boxing PASSED
numba_scipy/tests/test_sparse.py::test_sparse_shape PASSED
numba_scipy/tests/test_sparse.py::test_sparse_ndim PASSED
numba_scipy/tests/test_sparse.py::test_sparse_copy PASSED
```

**Comment by brandonwillard** [comment]
> All tests pass locally, but I had to manually import `sparse` at the top level of `test_sparse.py`

Fixed.

**Comment by guilhermeleobas** [comment]
@brandonwillard, code looks good. My only suggestion is to emit a `NumbaExperimentalFeatureWarning`. Also, it looks like you have a failure in one of the [Azure CI jobs](https://dev.azure.com/numba/numba-scipy/_build/results?buildId=14088&view=logs&j=22f2e658-3a30-5b6f-e6b1-9372eade2c06&t=51c629d2-7330-5827-1b39-aefbf311aed4).

As for the current `catchsegv` failure. Could you add the following to the github actions file:
```
- name: Install glibc-tools
  shell: bash -l {0}
  run: |
    if [[ $(uname) == Linux ]]; then
      sudo apt-get install glibc-tools
    fi
```

**Comment by brandonwillard** [comment]
> @brandonwillard, code looks good. My only suggestion is to emit a `NumbaExperimentalFeatureWarning`. 

Should this be on module load?  Seems like that might cause the warning to be emitted whenever `numba_scipy` is loaded, no?

> Also, it looks like you have a failure in one of the [Azure CI jobs](https://dev.azure.com/numba/numba-scipy/_build/results?buildId=14088&view=logs&j=22f2e658-3a30-5b6f-e6b1-9372eade2c06&t=51c629d2-7330-5827-1b39-aefbf311aed4).

Oh, perhaps that's due to the changes between Numba 0.47.0+0.g4eb9cf875.dirty and 0.53.1. 


> As for the current `catchsegv` failure. Could you add the following to the github actions file:

I added that liberally.

**Comment by brandonwillard** [comment]
Yes, that's it: Numba 0.47 appears to have some important differences that cause this approach to fail.

File: .github/workflows/test_and_deploy.yml [hunk 1]
```diff
@@ -99,6 +111,12 @@ jobs:
     env:
       AC_LABEL: dev
     steps:
+      - name: Install glibc-tools
+        shell: bash -l {0}
+        run: |
+          if [[ $(uname) == Linux ]]; then
+            sudo apt-get install glibc-tools
+          fi
```
Comment (guilhermeleobas) [inline .github/workflows/test_and_deploy.yml:119]: You don't need glibc in this job

File: .github/workflows/test_and_deploy.yml [hunk 2]
```diff
@@ -68,6 +74,12 @@ jobs:
         runs-on: [macos-latest, ubuntu-latest, windows-latest]
     runs-on: ${{ matrix.runs-on }}
     steps:
+      - name: Install glibc-tools
+        shell: bash -l {0}
+        run: |
+          if [[ $(uname) == Linux ]]; then
+            sudo apt-get install glibc-tools
+          fi
```
Comment (guilhermeleobas) [inline .github/workflows/test_and_deploy.yml:82]: You don't need glibc in this job

**Comment by guilhermeleobas** [comment]
> Yes, that's it: Numba 0.47 appears to have some important differences that cause this approach to fail.

I'm impressed that one of the CI jobs is using Numba 0.47 as it is almost ten versions older. I'll raise this at the next public meeting to update the CI config used here.

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]

**Comment by stuartarchibald** [comment]
> > > Yes, that's it: Numba 0.47 appears to have some important differences that cause this approach to fail.
> > 
> > 
> > I'm impressed that one of the CI jobs is using Numba 0.47 as it is almost ten versions older. I'll raise this at the next public meeting to update the CI config used here.
> 
> Thanks; making this support much older versions already seems like it could be a real hassle.

I suspect this is coming from CI using Python 3.6 and 3.7, Numba 0.57 (currently in development) has a baseline of Python 3.8. Perhaps move all the Python versions in CI up by a `.1` and see if that helps?

**Comment by guilhermeleobas** [comment]
@brandonwillard, can you add docs for your changes?

**Comment by guilhermeleobas** [comment]
@esc, can we merge this PR?

**Comment by brandonwillard** [comment]
> @brandonwillard, let me know once you have the chance to update the docs and we can merge this PR.

I'll try to do that in the next day or so.  If I don't, feel free to send me another reminder.

**Comment by guilhermeleobas** [comment]
Hi @brandonwillard. Quick update on my side. I'll review this PR later this week.

File: numba_scipy/sparse.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    def __init__(self, dtype):
+        warnings.warn(
+            "Use of sparse matrices detected. This is an experimental feature.",
+            category=NumbaExperimentalFeatureWarning,
+        )
+        self.dtype = dtype
+        self.data = types.Array(dtype, 1, "A")
+        self.indices = types.Array(types.int32, 1, "A")
+        self.indptr = types.Array(types.int32, 1, "A")
```
Comment (ivirshup) [inline numba_scipy/sparse.py:55]: Minor thing, but these can also be `int64`. Scipy is pretty aggressive about down casting it to int32, but if you have enough values to need 64 bit integers they will stick around.

Comment (brandonwillard) [reply numba_scipy/sparse.py:55]: > Minor thing, but these can also be `int64`. Scipy is pretty aggressive about down casting it to int32, but if you have enough values to need 64 bit integers they will stick around.
> 
> Super excited to see this moving forward!

Thanks for the heads up!  If no one objects, I'll be glad to change them.

**Review by guilhermeleobas — DISMISSED** [review]
Thanks, @brandonwillard.

**Comment by guilhermeleobas** [comment]
@esc, Could you take a look at this PR and merge it if possible?

File: numba_scipy/sparse.py (hunk 3 above)
Comment (ivirshup) [reply numba_scipy/sparse.py:55]: Linking in some relevant discussion from scipy, since it came up on my notifications again: https://github.com/scipy/scipy/issues/16774

tldr (from my perspective) scipy wants to be more supportive of int64 indices, and remove the automatic down casting to int32 since checking the values is expensive.

File: docs/source/reference/sparse.rst [hunk 4]
```diff
@@ -0,0 +1,8 @@
+Supported functions from ``scipy.sparse``
+=========================================
+
+Sparse matrix support is currently experimental and limited to boxing and
+unboxing of the :py:class:`scipy.sparse.csr_matrix` and
+:py:class:`scipy.sparse.csr_matrix` types and support for the
```
[the remark this answers is not in the document]
Comment (esc) [reply docs/source/reference/sparse.rst]: Thank you @stuartarchibald I committed the suggestion.

**Comment by esc** [comment]
@stuartarchibald thank you for the last review. I will merge this now trusting everyone involved. Thank you all for the patch!

**Review by esc — APPROVED** [review]
LGTM!

**Comment by esc** [comment]
I deployed a fresh token and retriggered the action to produce a new dev package:

https://anaconda.org/numba/numba-scipy/files?version=0.4.0dev0&type=conda&channel=dev
